Sprint 17

medvedevda opened the pull request:
Выполнил спринт 17. Очень понравился, интересный, много работы. Хотел бы услышать обратную связь. Также есть несколько вопросов: насколько нужно было внедрять в проект swiftGen? Я обошелся без него и сделал локализацию через enum. При редактировании трекера возникли сложности в установке новых элементов/правке констреинтов и изменении таргетов к кнопкам. Пришлось передавать через init параметры и вызывать методы раньше. Есть ли какие альтернативные варианты второго вопроса?

Commit:
Добавлена поддержка следующих языков: Русский, Английский, Иврит

datePicker устанавливается в зависимости от выбранного региона
локализированы все экраны приложения

Commit:
Добавлена плюрализация строк для всех языков. Кол-во дней в рекордах отображается корректно

Commit:
Добавил ограничение на кол-во вводимых символов в textFiled.

Теперь появляется предупреждение

Commit:
Ограничение по кол-ву символов в textField теперь появляется во время ввода

Commit:
Добавил локализацию для FiltersVC. Добавил алерт для подтверждения удаления трекера

Создал enum для заполнения всех таблиц в приложении

Commit:
Реализовано закрепление/открепление трекеров

исправлен цвет separator в tableViews
добавлена локализация для категории "Закрепленные"

Commit:
Написаны snapshot тесты для светлой и темной тем. Реализовано редактирование категории

Commit:
Добавлена возможность редактировать трекер и удалять категории

Commit:
Добавлена блокировка кнопки "Готово" в scheduleVC, если не выбрано расписание

при редактировании трекера в расписании отображаются ранее выбранные дни

Commit:
Подключена AppMetrica. Написаны функции для активации Яндекс.Метрики и отслеживания событий.

Добавлены enums для события/экранов/кнопок

Commit:
Добавлены отправки событий от действия пользователя для сбора аналитики на всех экранах

Commit:
fix

Commit:
Создал UI для StatisticVC с учетом показателей. Создал statisticViewModel

Commit:
Реализована расширенная статистика для трекеров

Commit:
Выполнен спринт 17

- добавлена темная тема
- реализована статистика трекеров
- добавлена аналитика действий пользователя
- реализован фильтрация трекеров в filterVC
- локализованы все экраны в проекте и переведены на 2 доп. языка
- реализовано контекстное меню на главном экране, с помощью которого можно закреплять/редактировать/удалять трекеры
- реализовано контекстное меню на экране категорий, с помощью которого можно редактировать/удалять категории

amardanov commented on Tracker/Controllers/Statisctic/StatisticCellView.swift:
```diff
@@ -0,0 +1,61 @@
+
+final class StatisticCellView: UIView {
+    
+    lazy var countLabel: UILabel = {
```
private должны быть все UI элементы

amardanov commented on Tracker/Controllers/Statisctic/StatisticView.swift:
```diff
@@ -0,0 +1,49 @@
+import UIKit
+
+final class StatisticView {
+    lazy var emptyLabel: UILabel = {
```
Тоже private UI элементы

amardanov commented on Tracker/Controllers/Trackers/CreateTracker/CreateTrackerView.swift:
```diff
@@ -10,7 +10,7 @@ import UIKit
 final class CreateTrackerView {
     lazy var createTrackerLabel: UILabel = {
```
private

amardanov commented on Tracker/Controllers/Trackers/CreateTracker/NewTracker/Category/EditingCategory/EditingCategoryView.swift:
```diff
@@ -0,0 +1,44 @@
+import UIKit
+
+final class EditingCategoryView {
+    lazy var titleLabel: UILabel = {
```
private

amardanov commented on Tracker/Controllers/Trackers/CreateTracker/NewTracker/Category/EditingCategory/EditingCategoryViewController.swift:
```diff
@@ -0,0 +1,100 @@
+}
+
+//MARK: SetupViews
+extension EditingCategoryViewController {
```
можно весь extension сделать private и не писать потом это перед каждым методом

amardanov commented on Tracker/Controllers/Trackers/CreateTracker/NewTracker/Category/NewCategory/NewCategoryView.swift:
```diff
@@ -11,20 +11,21 @@ final class NewCategoryView {
         return element
     }()
     
     lazy var newCategoryTextField: UITextField = {
```
private

amardanov commented on Tracker/Controllers/Trackers/CreateTracker/NewTracker/Category/NewCategory/NewCategoryViewController.swift:
```diff
@@ -10,17 +10,23 @@ import SnapKit
     }
     
+    deinit {
+        analyticsService.report(typeOfEvent: .click, screen: .newCategoryVC, item: nil)
```
вынеси в метод `viewDidDisappear`

amardanov commented on Tracker/Controllers/Trackers/CreateTracker/NewTracker/NewTrackerView.swift:
```diff
@@ -16,12 +16,21 @@ final class NewTrackerView {
         return element
     }()
     
+    lazy var countOfCompletedDaysLabel: UILabel = {
```
private

amardanov commented on Tracker/Controllers/Trackers/CreateTracker/NewTracker/Schedule/ScheduleView.swift:
```diff
@@ -10,7 +10,7 @@ import UIKit
 final class ScheduleView {
     lazy var scheduleLabel: UILabel = {
```
private

amardanov commented on Tracker/Controllers/Trackers/Filter/FilterTableViewCell.swift:
```diff
@@ -9,7 +9,7 @@ import UIKit
 final class FilterTableViewCell: UITableViewCell {
     
-    private lazy var label: UILabel = {
+    lazy var label: UILabel = {
```
private

amardanov commented on Tracker/Controllers/Trackers/Filter/FilterViewController.swift:
```diff
@@ -24,16 +22,25 @@ final class FilterViewController: UIViewController {
     }
     
+    deinit {
+        analyticsService.report(typeOfEvent: .close, screen: .filterVC, item: nil)
```
тоже лучше в метод `viewDidDisappear`

amardanov commented on Tracker/Controllers/Trackers/TrackersView.swift:
```diff
@@ -36,15 +36,15 @@ final class TrackersView {
         return element
     }()
     
     lazy var searchTextField: UISearchTextField = {
```
private

amardanov commented on Tracker/Controllers/Trackers/TrackersViewController.swift:
```diff
@@ -52,11 +69,11 @@ final class TrackersViewController: UIViewController, TrackerViewControllerProto
         if trackersViewModel.areVisibleCategoriesEmpty() {
```
`trackersViewModel.areVisibleCategoriesEmpty() ? addEmptyViews() : reloadCollectionView()`

amardanov commented:
Все сделано хорошо. SwiftGen не нужен тут, как ты сделал - это хорошо) 
Единственное, много лишнего попало в гит. Рекомендую почитать про gitignore, сам файлик можно найти [тут](https://github.com/github/gitignore/blob/main/Swift.gitignore)

Commit:
Исправлено название главного экрана в analyticsService

исправлена логика отображения категории "Закрепленные" на главном экране - теперь отображается в самом верху
каждое событие для отправки аналитики содержит верные три поля (event, screen, item)
отправка событий о закрытии контроллеров теперь реализуется во viewDidDisappear
прописаны пропущенные в некоторых местах self

Commit:
fix

Commit:
реализован метод делегата для scrollView:

чтобы скрывалась клавиатура и сохранялось новое название трекера

Commit:
fix bugs

Commit:
fix

Commit:
fix
